posthog-server/src/main/java/com/posthog/server/internal/PostHogFeatureFlags.kt

```diff
@@ -72,28 +241,249 @@ internal class PostHogFeatureFlags(
                 groupProperties = groupProperties,
             )
 
-        // Check cache first
         val cachedFlags = cache.get(cacheKey)
         if (cachedFlags != null) {
-            config.logger.log("Feature flags cache hit for distinctId: $distinctId")
             return cachedFlags
         }
 
-        // Cache miss
-        config.logger.log("Feature flags cache miss for distinctId: $distinctId")
         return try {
             val response = api.flags(distinctId, null, groups, personProperties, groupProperties)
             val flags = response?.flags
             cache.put(cacheKey, flags)
             flags
         } catch (e: Throwable) {
-            config.logger.log("Loading feature flags failed: $e")
+            config.logger.log("Loading remote feature flags failed: $e")
             null
         }
     }
 
+    override fun getFeatureFlags(
+        distinctId: String?,
+        groups: Map<String, String>?,
+        personProperties: Map<String, Any?>?,
+        groupProperties: Map<String, Map<String, Any?>>?,
+    ): Map<String, FeatureFlag>? {
+        if (distinctId == null) {
+            config.logger.log("getFeatureFlags called but no distinctId available for API call")
+            return null
+        }
+
+        val cached = getFeatureFlagsFromCache(distinctId, groups, personProperties, groupProperties)
+        if (cached != null) {
+            return cached
+        }
+
+        // If no cached flags, try local evaluation
+        val localFlags =
+            getFeatureFlagsFromLocalEvaluation(
+                distinctId,
+                groups,
+                personProperties,
+                groupProperties,
+            )
+        if (localFlags != null) {
+            return localFlags
+        }
+
+        // Finally, fall back to remote fetch
+        return getFeatureFlagsFromRemote(distinctId, groups, personProperties, groupProperties)
+    }
+
     override fun clear() {
         cache.clear()
         config.logger.log("Feature flags cache cleared")
     }
+
+    override fun shutDown() {
+        stopPoller()
+    }
+
+    /**
+     * Load feature flag definitions from the API for local evaluation
+     */
+    public fun loadFeatureFlagDefinitions() {
+        if (!localEvaluation || personalApiKey == null) {
+            return
+        }
+
+        var wasWaitingForLoad = false
+
+        synchronized(loadLock) {
+            while (isLoading) {
+                wasWaitingForLoad = true
+                try {
+                    loadLock.wait()
+                } catch (e: InterruptedException) {
+                    Thread.currentThread().interrupt()
+                    config.logger.log("Interrupted while waiting for flag definitions to load")
+                    return
```

Comment:
This returns without notifying waiters afaict, which means that if it's interrupted, it'll return without setting `isLoading = false`, so other threads might wait.  

Maybe do something like this?

```
synchronized(loadLock) {
    while (isLoading) {
        try {
            loadLock.wait()
        } catch (e: InterruptedException) {
            Thread.currentThread().interrupt()
            isLoading = false
            loadLock.notifyAll()
            throw InterruptedException("Interrupted while waiting")
        }
    }
}
```